### ISSUE TITLE
No completion support for Null<T>

### ISSUE THREAD

**Issue by winmain** [HaxeFoundation/intellij-haxe#516]
I want to use `Null<T>` construction when I want to annotate some method that can return null. Sadly, completion for that method will be broken.

![Example](https://cloud.githubusercontent.com/assets/167248/19879639/d528577a-a02f-11e6-8a96-a6b87e28b494.png)

For copy-paste:
```haxe
class Test {
  public function foo(): Null<String> return "";

  public function test() {
    foo().| // <- no suggestions
  }
}
```
Plugin version: `0.10.1-RC2`
Idea: `2016.2`

**Comment by EBatTiVo** [comment]
I wonder if this is broken in general, or only with classes that use inline preprocessor directives (e.g. #if..#endif).  String, in particular, uses them extensively and won't parse correctly as a consequence.  (See bug #115).

**Comment by winmain** [comment]
I think use preprocessor directives doesn't matter here. And my small example does not fully reproduce strange behavior. I will make better example.

**Comment by winmain** [comment]
File `Test.hx`:
```haxe
class Test {
  public function foo(): Null<Foo> return "";

  public function test() {
    foo().|
  }
}

class Foo {
  public var a: Int;
}
```

File `T.hx`:
```haxe
class T {
  public var tt: Int;
}
```

In this example return type of method `foo()` incorrectly treated as `T`.
The important part here is a `T` class in **separate** file. If I rename it or move in `Test.hx` the bug disappears.

I think somewhere deep in the code name `T` hardcoded as template name or something like. So if you have a class named `T` in a project, you will face to strange bugs :)

[example.zip](https://github.com/TiVo/intellij-haxe/files/644991/example.zip)

**Comment by EricBishton** [comment]
Verified.  Still exists in  0.11.0.RC-3.

**Comment by EricBishton** [comment]
The name T is used in the Null<T> definition. (Not that there aren't any hard-coded 'T' strings in the code, but they are now gone from the resolver.)  I think the problem here is that the HaxeResolveUtil.resolve() tries to look for a class before it looks in the specialization for the name T.  This might be a simple ordering issue.

**Comment by winmain** [comment]
Did you mean HaxeResolver.resolve()?

**Comment by EricBishton** [comment]
I think I meant HaxeClassResolveUtil.getHaxeClassResolveResult().  But it's one or the other of those classes.

**Comment by EricBishton** [comment]
Works in 1.1 release (or, at least, 7329edcd).  Not sure which change fixed it - probably merged in with 3ea6747d3.
